### PR TITLE
BytesToBGRImg to support resizing

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

1) Refinement to BytesToBGRImg to support resize, no resizing by default to make it backward compatible 
2) Load ImageNet mean from txt file and create Tensor instead of Tensor object
## How was this patch tested?

Tested locally with Alexnet validation

### CODE REVIEW

**Comment by jason-dai** [comment]
How about add a Resize transformer and then apply BytesToBGRImg->Resize?

**Comment by wzhongyuan** [comment]
@jason-dai  I did try doing it, but in order to leverage java.awt.image to draw a resized image, we need to pass in the original image content as a byte array, which means after BytesToBGTImg convert bytes to float, we need to convert back to bytes again, and it performs very badly as we had to copy byte one by one as follow

 val bufferedImage : BufferedImage
        = new BufferedImage(width, height, BufferedImage.TYPE_3BYTE_BGR)
      val outputImagePixelData = bufferedImage.getRaster.getDataBuffer
        .asInstanceOf[DataBufferByte].getData
      val imgContent = img.content.map(_.toByte) //float to byte conversion
      var i = 0
      while (i < width * height * 3) {
        outputImagePixelData(i) = imgContent(i)
      }

Maybe we leave this transformer for now ?

**Comment by jason-dai** [comment]
Maybe you can check if this can be supported in the new `transform` package in zoo: https://github.com/intel-analytics/analytics-zoo/pull/23

**Comment by wzhongyuan** [comment]
@jason-dai  sure, I will check it out, thanks

**Comment by wzhongyuan** [comment]
@jason-dai    As discussed offline, I've made refinement to BytesToBGRImg transformer to support resizing and be backward compatible 
Tested locally with model validations
please help review and comment

**Comment by wzhongyuan** [comment]
PR validation passed http://10.239.47.211:18888/job/BigDL-PR-Validation/1751/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/image/BytesToBGRImg.scala [hunk 1]
```diff
@@ -16,27 +16,78 @@
...
+      val outputImagePixelData = bufferedImage.getRaster.getDataBuffer
+        .asInstanceOf[DataBufferByte].getData
+      System.arraycopy(imgBuffer.array(), 8,
+        outputImagePixelData, 0, outputImagePixelData.length)
+      resizeImage(bufferedImage, resizeW, resizeH)
+    }
+  }
+
+  private def resizeImage(img: BufferedImage, resizeWidth: Int, resizeHeight: Int): Array[Byte] = {
+    var scaledImage: java.awt.Image = null
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/image/BytesToBGRImg.scala]: why not reuse the method  BGRImage.resizeImage

**Comment by wzhongyuan** [comment]
@SeaOfOcean  Thanks for reminder, didn't realize its exposed. Refined for review

**Comment by wzhongyuan** [comment]
@SeaOfOcean  Latest change has been tested as well

**Comment by wzhongyuan** [comment]
@SeaOfOcean   Please help review this PR as well, PR validation passed 
 http://10.239.47.211:18888/job/BigDL-PR-Validation/1846/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala [hunk 2]
```diff
@@ -32,24 +33,35 @@ object AlexNetPreprocessor {
 
   def apply(path: String, batchSize: Int, meanFile: String, sc: SparkContext)
   : DataSet[MiniBatch[Float]] = {
-    val means = File.load[Tensor[Float]](meanFile)
+    val means = createMeans(meanFile)
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala]: Add comments to explain the format of meanFile? And may need provide a download link to get the meanFile in the readme

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala [hunk 3]
```diff
@@ -32,24 +33,35 @@ object AlexNetPreprocessor {
...
   }
+
+  private def createMeans(meanFile : String) : Tensor[Float] = {
+    val lines = Files.readAllLines(Paths.get(meanFile), StandardCharsets.UTF_8)
+    val array = new Array[Float](lines.size)
+    lines.toArray.zipWithIndex.foreach {
+      x => {
+        array(x._2) = x._1.toString.toFloat
+      }
+    }
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala]: val array = Source.from(meanFile).getLines().map(_.toFloat) ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala (hunk 2 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala]: Comments added, User should have option to convert from imagenet mean file as long as we provide the format, we could upload a preconverted one as well

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala (hunk 3 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala]: Much cleaner, thanks

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala [hunk 4]
```diff
@@ -26,30 +27,39 @@ import com.intel.analytics.bigdl.utils.File
...
       // do not normalize the pixel values to [0, 1]
-    val transfomer = BytesToBGRImg(normalize = 1f) ->
+    val transfomer = BytesToBGRImg(normalize = 1f, 256, 256) ->
       BGRImgPixelNormalizer(means) -> BGRImgCropper(imageSize, imageSize, CropCenter) ->
       BGRImgToSample(toRGB = false)
     transfomer(dataSet)
   }
+
+  def createMeans(meanFile : String) : Tensor[Float] = {
+    val array = Source.fromFile("/home/jerry/Downloads/mean.txt").getLines().map(_.toFloat).toArray
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala]: don't use the path here

**Comment by wzhongyuan** [comment]
PR validation passed with latest refinement

http://10.239.47.211:18888/job/BigDL-PR-Validation/1876/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala [hunk 5]
```diff
@@ -26,30 +27,39 @@ import com.intel.analytics.bigdl.utils.File
...
+
 
 object AlexNetPreprocessor {
   val imageSize = 227
 
   def apply(path: String, batchSize: Int, meanFile: String, sc: SparkContext)
   : DataSet[MiniBatch[Float]] = {
-    val means = File.load[Tensor[Float]](meanFile)
+    // 'meanFile' specify the path to the pixel level mean data, one line per pixel
+    // following N * H * W * C order, 196608 in total (256 * 256 * 3)
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala]: There is no batch => H * W * C

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/loadmodel/DatasetUtil.scala]: @SeaOfOcean  thanks, fixed typo~

**Comment by wzhongyuan** [comment]
PR validation passed http://10.239.47.211:18888/job/BigDL-PR-Validation/2033/  
@SeaOfOcean  As discussed, closing this one

**Pull request merged**